succ

ManofSausage opened the pull request:
### Ling 2.0, Ling Rework, NuLing, ReenLing, Abomination that haunts my dreams oh my god how many hours have you stolen from me that I should be touching grass.

**Acknowledgements and Special Thanks**
I know you normally put this thing at the end, but after everything these boys have done for the project, I feel like this has to go front and center. Thank you sm @Ryan180602, who first had the idea for this project, did pretty much all the real work I added onto, and without whom this project wouldn't exist at all. You're a real one. The second thanks goes to @SomeAngryMiner whose sprites gave this rework all of its personality. Having sprites that good straight up kept me from giving up on the project altogether sometimes, and it straight up wouldn't even have made it to draft PR if not for you.

**At long last, after relentlessly promising, it's FINALLY up as a draft PR. The draft still has a number of issues that'll need to be resolved before I change this to a real PR though/this is worthy of review.**

-Lings do not copy biomods properly, carrying tattoos/sprite_accessories across transformations.
-Basically all of the spell icons are still old tg icons that look ugly, will replace with Azlan sprites

This is going to have a MASSIVE changelog by the end of this. Since it's still a draft, I'll have an additions section for now, that explains how this port is differentiated from current ling.
### Removals:
**Where is lesser form? Is he safe? Is he alright?**
I killed him. I ate his spleen and deleted him. Lesser form is infamously hated by bay players, and we all know why. It reduces what should be a horrific mode to monke and turns sec into Petrov babysitters.

### Additions:
**Spells Menu** Instead of selecting and using powers from a messy side tab, lings will now be able to use their unique fleshy powers from a spells menu at the top of the screen.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/45492628-bd80-4d40-b69e-7361b98986ff)

The main offering of new and yassified ling, of course, is it's new and yassified set of powers. Worth keeping in mind is that a single ling will only be able to take a fraction of these powers in a single round, forcing lings to specialise and pick their powers carefully. The intent is that lings will be both 1) Balanced, and incapable of becoming unstoppable murder machines by having every single power, and 2) Encouraged to experiment with different combinations of powers in round. Lings will spawn with 12 evo points at roundstart.

**RESURRECTION** While I will not spoil this(except to those of you who want to code dive), Ling will come with a new special resurrection mechanic with implications for what kind of creature they are. No longer will merely cutting the head off be enough. With proper preparation, lings might even be able to survive the destruction of their own body.

**Arm Hammer (2):** Reforms the ling's arm into a giant meat and bone hammer. Mid to pretty good damage, but extremely good armor penetration especially once upgraded. Gets a smash attack on disarm which can send opponents flying, and is great for avoiding getting pinned.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/ab271f6a-7770-47d3-818c-5040386babaf)
**Claw (1):** Forms the ling's arm into a fearsome claw. Mid damage, but reliable armor pen and can be upgraded with recursive enhancement. Gets a grab attack on grab intent which can instantly aggo grab enemies in melee range
![image](https://github.com/Baystation12/Baystation12/assets/134344460/dbbdd439-f953-4389-bf7c-17dad30faf99)
**Arm Shield (2):** Forms the ling's arm into a powerful shield. No special ability, but is as strong as a riot shield, and can be upgraded to be as strong as a plasteel riot shield. Capable of deflecting lasers in it's upgraded form.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/142a2db8-4698-4544-bc33-bf1b34bc58fa)
**Organic Space Suit (1):** Forms a fleshy spacesuit around the ling's body. Voidproof, conceals the identity of the wearer and confers absolutely zero slowdown, allowing the ling to move at peak speed in it. However, it provides no armor.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/68bd59fb-a41f-4c72-81bd-81f86d9aecb6)
**Arm Blade (2):** The famous big one y'all were waiting for. Incredibly damaging even in its unupgraded form, it's upgraded form is only a step below esword tier, with brutal damage, armor penetration and parry chance. Has the ability to rip open airlocks and firelocks, taking longer depending on the strength of the airlock.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/bf2ed9c3-f3ba-41f0-8ba1-71929d06e1ad)
**Chitinous Spacearmor (3):** RIP AND TEAR. Forges an incredibly powerful suit of monstrous armor around the ling's body that is not only spaceproof but confers NO slowdown. However, the armor comes with a catch: It has weak resistance to lasers and fire, and the ling cannot used ranged weapons at all while it is active.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/17fb987f-ded9-40fa-a3c2-6ee7ded91846)
**Augmented Eyesight (2):** Grows the ling a set of organic lenses, which upon activating this power, provide the ling with thermal vision for five seconds before their sight returns to normal. Only really useful for pulses, and too chemical intensive to activate constantly.
**Bioelectrogenesis (2):** Charges the ling's hand with bioelectricity, allowing them to shock victims by touching them, or grabbing them as well as recharging items. Grabbed victims are hurt more, and receive a larger shock.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/ffa97894-2ad9-4306-96f8-6fc63cca6b6a)
**Boost Range(1):** Allows the ling to grow a single shot dart gun, which selects a type of sting upon being grown. Upon being hit by the dart, organics will be affected with the effect of the sting. Blocked by any thick covering that blocks the regular dart gun.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/d66a9f19-c4aa-4993-a866-388c22185a82)
**Blind Sting (2):** Does exactly what it did in old ling, temporarily blinds the victim.
**Death Sting (7):** Does exactly what it did in old ling, but the cost has been kept massively expensive to avoid it being abused. May just remove this entirely.
**Cryogenic Sting (1):** Injects the victim with cryoxadone, slowing them and lowering their temperature, causing damage from the extreme cold. Injects more on recursive enhancement.
**Dark Sight (default):** Default roundstart ability. Gives the ling night vision on command.
**Deaf Sting (1):** Same as in old ling. Deafens longer with recursive enhancement.
**Delayed Toxic Sting (1):** After two minutes of being stung, the victim will suffer minor toxic damage, causing them to vomit. Essentially useless as a killing agent, but VERY useful at getting a guard or friend to go to medical to ambush someone alone.
**Electric Lockpick (3):** Reshapes a finger into being capable of carrying an electric charge that can open doors, and only doors. The breaching process takes time, and will cost chemicals per door unlocked.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/364ed4b4-7cf2-4315-a192-6fc44df6ebad)
**Endoarmor (1):** Grows a modified form of subdermal armor under the ling's chest which provides partial protection from brute damage, but does nothing whatsoever to protect from burns.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/53070bb7-607f-4780-bcfb-da1ff5cc2302)
**Engorged Chemical Glands(1):** You get more chems. What it says.
**Enrage (2):** Grows a sac that releases adrenaline and painkilling chemicals upon command. The ling will temporarily become almost immune to pain and superhumanly fast, but only for a short time. After the effect ends, the ling will be massively weakened and slowed, and may even pass out. Lasts longer with recursive enhancement.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/6343e783-35fc-4856-9d04-3a065711c2b6)
**Extract DNA Sting(default)** Same as old ling. You eat-a their dna(but just a nibble frfr), and can now copy other people's form
**Escape Restraints (2):** Allows the ling to escape from cuffs, legcuffs and other restraints, same as /tg/. Cooldown is shorter with recursive enhancement.
**False Identity (1):** Allows the ling to create custom disguises similar to the facial surgeon item, transforming their name, appearance and even gender.
**Fabricate Clothing(1):** Allows the ling to grow a biological camo kit, including the ability to secrete accessories at will. Become the perfect duplicate. Steal their identity flawlessly.
**Fleshmend (1):** Quick surface level healing that provides no warning to onlookers. Allows the ling to quickly heal from superficial damage, but does nothing to heal organs, arterials or more serious damage. Disabled by being on fire. Stronger healing with recursive enhancement.
**Hallucination Sting(1):** Same as in current. You trip balls.
**Mimic Voice(1):** Same as in current. You fake-a the voice!
**Anatomic Panacea(1):** Purges all poisons, chems and radiation from the ling's body. Avoid the consequences of your crippling chloral addiction.
**Paralysis Sting(3):** Temporarily paralyses the victim, dropping them similar to Vecuronium. VERY on the fence about this one, on the one hand ling needs an ambush method on the other hand I hate parapen.
**Rapid Regeneration(2):** The ling begins to regenerate slowly, and after 10 seconds will heal a large amount of damage of every type, one external bodypart with all it's organs and spew blood and gibs everywhere. This process requires the ling to stand still the entire time, and will be interrupted if they move or attack. Disabled by being on fire. Stronger healing with recursive enhancement.
**Rapid Chemical Synthesis(1):** You generate 1 chem per second, instead of 0.5
**Recursive Enhancement(3):** Gives very strong buffs to the next ability you activate, described under the descriptions of specific abilities.
**Regrow Limb(1):** What it says, you regrow a single bodypart but at huge chemical cost, making this a very inefficient way to heal the entire body. Cannot heal the torso or head.
**Reinforced Tendons(1):** You unlock the instant leap grab jump as a maneuver. May add a resistance to fall damage to this, if so will increase cost.
**Self Respiration(default):** Comes unlocked roundstart. Breathing is for chumps, allows you to have no need to breathe on being toggled.
**Resonant Shriek(2):** Release a powerful shriek that blows lights, deafens people and shuts off flashlights. You live in da shadows. Range is better with recursive enhancement.
**Dissonant Shriek(2):** Release a powerful shriek that shuts off radios in an area around you. Let them try to call for help now. Range is better with recursive enhancement.
**Silence Sting(2):** Same as in current ling, mutes the victim for some time, longer with recursive enhancement.
**Synaptizine Overdose(2):** Removes any stuns, wakes the ling from unconsciousness if they've been knocked out and basically gets you back on your feet. Does NOT remove chems from the body, or undo any damage the ling has sustained.
**Tentacle(2):** Grow a tentacle capable of grabbing items from a distance(GIMME THAT) or pulling enemies towards you, as well as weakening them and causing stamina damage. Get back here you little shit!
**Transform(default):** Default ability. Take their face! Impersonate them! Steal their skins! YOU ARE THE SKINWALKER.
**Camouflage (3):** Makes the ling partially but not quite transparent, allowing them to be spotted by a perceptive eye. Halts chemical regen while active, and forces the ling to move slowly. Attacking or interacting with items will end camo, as will running without recursive enhancement.
![image](https://github.com/Baystation12/Baystation12/assets/134344460/58a3cae2-71c2-430f-ae7a-7a14a662cef6)

Ryan180602 committed:
hell (changeling port polaris)

SierraKomodo reviewed (dismissed):
You have a lot of compilation warnings and test failures.

ManofSausage commented:
> You have a lot of compilation warnings and test failures.

Yeeee sorry just rebased so tests will be shidding themselves even more than normal, I'll fix them and bring it out of draft.

Is still waiting on a few sprites but I can do that before they come in.

ManofSausage commented:
@SierraKomodo Moving this out of draft since the only things missing now are 
1) Tentacle sprites, which are non essential for first release/test round. If it comes to it, I can just cut it for a test round and add them whenever the sprites are ready
2) A single failing test due to spawn cases(New ling has less spawn() uses since old ling had a bunch of them). Is there a way to get around that check since it'd feel goofy to have to add spawn() to match up with the test count since having more is undesirable.

SierraKomodo reviewed (dismissed):
Test failure:

> FAIL 334 spawn uses (expecting exactly 338)

If you're removing `spawn()` calls, update `test\check-paths.sh` to reflect the new total count.

If you're _adding_ `spawn()` calls, don't unless you have a very good reason. 99.9% of cases, you can either use timers (Via `addtimer()`) or run a proc with `set waitfor = FALSE` instead.

ManofSausage commented:
The despawnification(desloppification) is done at last, removed any spawns so we actually have less of those now than pre PR.

Updated the .sh file so the tests'd match.

ManofSausage commented:
webasing again ÚwÚ

MuckerMayhem commented on code/datums/managed_browsers/changelingevolution.dm:
```diff
@@ -0,0 +1,166 @@
+			return
+
+		if(!Thepower)
+			to_chat(M.current, "Purchase failed. Inform a dev of this error.")
```
This would be better if it gave a stack trace, maybe replace it with a `CRASH()` instead of just a generic return message. Same with the error below.

MuckerMayhem commented on code/game/gamemodes/changeling/changeling_evolutiontree.dm:
```diff
@@ -0,0 +1,362 @@
+	var/power_category = null // _defines/gamemode.dm
+	var/sting_effect = /mob/proc/generic_sting // only override for sting powers, used for boost range
+	var/sting_duration = null
+	var/reagents_transferred = null
```
Comments should be of the `///` variant in this case and above the line they're describing.

MuckerMayhem commented on code/game/gamemodes/changeling/changeling_evolutiontree.dm:
```diff
@@ -0,0 +1,362 @@
+
+
+	if(!Thepower)
+		to_chat(M.current, "This is awkward.  Changeling power purchase failed, please report this bug to a coder!")
```
Do a `CRASH()` here.

SierraKomodo replied:
Has this one been addressed yet?

Ryan180602 replied:
Has been.

MuckerMayhem commented on code/game/gamemodes/changeling/changeling_evolutiontree.dm:
```diff
@@ -0,0 +1,362 @@
+		return
+
+	if(Thepower in purchased_powers)
+		//to_chat(M.current, "We have already evolved this ability!")
```
Remove or uncomment

MuckerMayhem commented on code/game/gamemodes/changeling/generic_equip_procs.dm:
```diff
@@ -0,0 +1,295 @@
+//This is a generic proc that should be called by other ling armor procs to equip them.
```
```suggestion
///This is a generic proc that should be called by other ling armor procs to equip them.
```

MuckerMayhem commented on code/game/gamemodes/changeling/generic_equip_procs.dm:
```diff
@@ -0,0 +1,295 @@
+
+	var/mob/living/carbon/human/M = src
+
+	var/success = 0
```
Use `TRUE/FALSE` instead of `0/1`

MuckerMayhem commented on code/game/gamemodes/changeling/generic_equip_procs.dm:
```diff
@@ -0,0 +1,295 @@
+		src.update_icon()
+		return success
+
+//This is a generic proc that should be called by other ling weapon procs to equip them.
```
```suggestion
///This is a generic proc that should be called by other ling weapon procs to equip them.
```

MuckerMayhem commented on code/game/gamemodes/changeling/generic_equip_procs.dm:
```diff
@@ -0,0 +1,295 @@
+		to_chat(src, SPAN_WARNING("We cannot shape a missing limb into a weapon!"))
+		qdel(W)
+		return 0
+	//src.put_in_hands(W)
```
```suggestion
```

MuckerMayhem commented on code/game/gamemodes/changeling/powers/absorb.dm:
```diff
@@ -0,0 +1,133 @@
+
+//Absorbs the victim's DNA. Requires a strong grip on the victim.
+//Doesn't cost anything as it's the most basic ability.
+/mob/proc/toggle_absorb_type()
```
```suggestion
/**
 * Absorbs the victim's DNA. Requires a strong grip on the victim.
 * Doesn't cost anything as it's the most basic ability.
 */
/mob/proc/toggle_absorb_type()
```

MuckerMayhem commented on code/game/gamemodes/changeling/powers/armblade.dm:
```diff
@@ -0,0 +1,343 @@
+			SPAN_WARNING("Our arm twists and mutates, transforming it into a deadly weapon."),
+			SPAN_ITALIC("You hear organic matter ripping and tearing!")
+		)
+		*/
```
Was this meant to be uncommented?

MuckerMayhem reviewed (commented):
From a surface level view, theres quite a few leftover commented out chunks and lines that should be cleaned up (many of which I didn't pick out in this review), and a few spots proper commenting could be done. I noticed a lot of proc comments use `//` instead of `///` or `/** * */`, which would also be good to fix.

MuckerMayhem reviewed (dismissed):
see above.

SierraKomodo commented:
@MuckerMayhem There's a second tag for test merge stuff so we can keep track of these :P

Ryan180602 commented:
Just waiting to see the icon-diff checks catch anything awry during merge conflict resolution. Else this should b gud

Ryan180602 commented:
So icon-diff bot never finished the checks on this :(

MuckerMayhem reviewed (approved):
Tentative approval, there's still a lot of leftover commented out code that it would be nice to not have, but I figure its more of a nitpick at this point.
